currency/translation.go

```diff
@@ -20,3 +20,83 @@ var translations = map[*Item]Code{
 	XDG.Item:  DOGE,
 	USDT.Item: USD,
 }
+
+// NewTranslations returns a new translation map, the key indicates the exchange
+// representation and the value indicates the internal representation/common/standard
+// representation. e.g. XBT as key and BTC as value, this is useful for exchanges
+// that use different naming conventions.
+func NewTranslations(t map[Code]Code) Translations {
+	lookup := make(map[*Item]Code)
+	for k, v := range t {
+		lookup[k.Item] = v
+	}
+	return lookup
+}
+
+// Translations is a map of translations for a specific exchange implementation
+type Translations map[*Item]Code
+
+// Translate returns the translated currency code, usually used to convert
+// exchange specific currency codes to common currency codes. If no translation
+// is found it will return the original currency code.
+func (t Translations) Translate(incoming Code) Code {
+	if len(t) == 0 {
+		return incoming
+	}
+	val, ok := (t)[incoming.Item]
+	if !ok {
+		return incoming
+	}
+	return val
+}
+
+// Translator is an interface for translating currency codes
+type Translator interface {
+	// TODO: Add a asset.Item param so that we can translate for asset
+	// permutations. Also return error.
+	Translate(Code) Code
+}
+
+// PairsWithTranslation is a pair list with a translator for a specific exchange.
+type PairsWithTranslation struct {
+	Pairs      Pairs
+	Translator Translator
+}
+
+// keyPair defines an immutable pair for lookup purposes
+type keyPair struct {
+	Base  *Item
+	Quote *Item
+}
+
+// FindMatchingPairsBetween returns all pairs that match the incoming pairs.
+// Translator is used to convert exchange specific currency codes to common
+// currency codes used in lookup process. The pairs are not modified. So that
+// the original pairs are returned for deployment to the specific exchange.
+// NOTE: Translator is optional and can be nil. Translator can be obtained from
+// the exchange implementation by calling Base() method and accessing Features
+// and Translation fields.
+func FindMatchingPairsBetween(this, that PairsWithTranslation) map[Pair]Pair {
```

Comment:
Only clunk I can see with this is that it doesn't offer the ability to backtranslate. You need to call this function twice to create two maps to do so. eg you cannot do this without two calls `FindMatchingPairsBetween` with the params in opposite order XBT-USDCM -> BTC-USDC -> XBT-USDCM

Its not a required change, but I think its some user-friendliness to add both translation keys in there. Its only one thing to call and hold for a user wishing to translate between things

Reply:
Yeah I can add that functionality and split `Translate` method to `TranslateToCommon` and `TranslateToExchange` or add a TODO and open another PR when it's actually being used in the library.

Reply:
I didn't realise this was waiting on my response. Any of those is fine